### ISSUE TITLE
Need a "transaction" or "lock" mechanism

### ISSUE THREAD

**Issue by teo1978** [krakjoe/apcu#142]
Consider this, very common, use scenario:

- I check if a cached value exists (either via apc_exists() or by directly trying to fetch it)
- if it does, I use it
- if it doesn't, I build it (e.g. after doing some queries to a database) and I cache it.

Seems right, doesn't it?
Except you **CAN'T DO THAT** in a concurrent scenario, that is, if more than one instance of the same PHP script may be doing the same at the same time. Which is almost always the case.

Otherwise, it may happen (and it often happens if the workload is high) that more than one instance of the same script, executing the above pseudocode concurrently, don't find the cached value, they all build it (so far so good), and they all try to store it, while only one should.

The **only** true solution to that (leaving alone euristic and unreliable workarounds) is to be able to do the abovementioned sequence of actions **atomically**, which is something there's currently no way to do with APCu.

Since there's no way you can write the above in a single call to a single APC function, the only way the above can be done atomically is through some **transaction** mechanism, that is, adding some function like `apc_start_transaction()` and `apc_end_transaction()`, which would guarantee atomicity of the execution of code inbetween, just like transactions in databases. Or a locking mechanism, such as `apc_acquire_lock($whatever)` and `apc_release_lock($whatever)`.

It's unbelievable that there's no such thing, as without it, a caching system (or any other data storage system for that matter) is effectively useless, as it is inherently condemn to become unstable under high load.

**Comment by krakjoe** [comment]
I'm extremely aware of this, I've seen it happen, it's been on my mind for more than a year ...

Allowing userland code to acquire a mutex is extremely dangerous, if the code errors out, we can't release it, and bring down the server ...

So, I've thought about introducing a new API before ... I went into some detail about that recently [here](https://www.reddit.com/r/PHP/comments/3hy051/caching_best_practice_should_i_always_check_if/).

I'm not sure what the solution is actually going to be to this, I have ended up bringing APCu to PHP7, but what I really want I think is a new cache ...

I have no time, I'm stretched extremely thin at the moment because of php7, I don't know when I'll have time to work on this, or if I'll be able too.

**Comment by teo1978** [comment]
I am interested in this:
> Now, stampede protection can help a little, but in the wrong place. Stampede protection in the APC(u) sense of the word doesn't actually stop many processes from generating the data for the cache, it **only stops them writing to the cache**, but the laborious thing that you want to avoid is still happening in many processes at once.

I understand the limitations of that, but it would already be a serious improvement if I could avoid processes other than the first to get blocked updating the cache (that is, many processes would't find the cached data; all of them, unfortunately, would generate it, but only one would attempt to update it). I fail to see how your second example accomplishes that.

**Comment by teo1978** [comment]
I don't get this part:
> Allowing userland code to acquire a mutex is extremely dangerous, **if the code errors out**, we can't release it, and bring down the server ...

Do you mean if the *user* code errors out? It should be easy to implicitly release all the locks in that case. There are a number of other PHP APIs that do that. If you acquire an flock, for example, and don't release it, it is automatically released at the end of the PHP script. If you begin a transaction or lock tables or whatever with a MySQL server, everything you locked is released with the connection with the server is closed.

Unless you mean if *your* code errors out. Then well, yes of course. If it's not rock-solid it may be disastrous. But that applies to any kind of API design you may come up with.

**Comment by krakjoe** [comment]
https://github.com/krakjoe/apcu/commit/03d186102b5409a923b5b0b6ac2a03ca8a4fdce2

The commit above introduces the api I described before, in the case when locking being used supports it ...

I'll maybe try to open it up to other kinds of locking ... needs testing ...

No plan to backport this for 5, someone else can, if they like ...

**Comment by krakjoe** [comment]
Just to clarify:

> they all try to store it, while only one should.

Only one will succeed, this is what stampede protection in APCu protects against.

**Comment by krakjoe** [comment]
Other kinds of locking should work now, pending testing ...

We should better define what supports recursion natively ...

**Comment by krakjoe** [comment]
Closing ...

**Comment by teo1978** [comment]
> Just to clarify:
> 
> > they all try to store it, while only one should.
> Only one will succeed, this is what stampede protection in APCu protects against.

We are talking about *add*, right? (as opposed to *set*), which only writes a value if it doesn't already exist. 
Only one will succeed, but how long will the others block waiting before they "find out" that the value is already being written and fail? I ask because:
- if they have to wait until the one which succeeds is finished writing and releases the lock, then having many concurrent *adds* of which only one succeeds would be *almost* as inefficient as having many concurrent *set*s - or well may be not that much, but considerably more costly than:
- if, instead, they have to wait only until they "realize that somebody is writing", then having many concurrent *add*s (of which only one succeeds) would be *almost* as efficient as having only the one necessary add (or very few more) (modulo the overhead of the request, of course).

Which one is it? I think the second should be possible, but I have just observed that when having a high load of unnecessary *add*s, many of which were unnecessarily *add*ing already existing values (due to a bug in the application code), the response time of the *add*s was very slow: I did expect a performance improvement by removing the bug causing the high quantity of unnecessary *add*s, but I expect the performance improvement to be barely noticeable, expecting that the unnecessary *add*s (attempting and failing to write already existing values) would have a non significant impact.
